### DISCUSSION TITLE
[GitHub Action] How to concatenate a string with a GitHub Actions environment variable?

### DISCUSSION THREAD

**Discussion by kevinadhiguna** [community/community#22548] [New to GitHub]
<p>Hi, I would like to ask how to concatenate a string and a GitHub Actions environment variable which is  <code>GITHUB_SHA</code> in this case.</p>
<p>List of GitHub Action environment variables : <a href="https://docs.github.com/en/actions/learn-github-actions/environment-variables" class="inline-onebox" rel="noopener nofollow ugc">Environment variables - GitHub Docs</a></p>
<p>This is the file (main.yml) :</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/4/5/45e3da8dbd2f9c108dfae963f1edd5e33131adab.png" data-download-href="/uploads/short-url/9Yh7qzurwq65RXaWI5QEiggg6HF.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/43397636/181071760-78c7c1cc-159d-4f67-93e9-433c88d52ef6.png" alt="image" data-base62-sha1="9Yh7qzurwq65RXaWI5QEiggg6HF" width="690" height="311" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/5/45e3da8dbd2f9c108dfae963f1edd5e33131adab_2_690x311.png, https://aws1.discourse-cdn.com/github/optimized/3X/4/5/45e3da8dbd2f9c108dfae963f1edd5e33131adab_2_1035x466.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/5/45e3da8dbd2f9c108dfae963f1edd5e33131adab.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/5/45e3da8dbd2f9c108dfae963f1edd5e33131adab_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1062×480 59.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>This is the error I got in GitHub Action :</p>
<pre><code class="lang-auto">error: invalid tag "***/strapi-v4-dockerize:development-$GITHUB_SHA": invalid reference format
Error: buildx failed with: error: invalid tag "***/strapi-v4-dockerize:development-$GITHUB_SHA": invalid reference format
</code></pre>
<p>Expected output :<br>
<code>myaccount/strapi-v4-dockerize:development-ffac537e6cbbf934b08745a378932722df287a53</code></p>
<p>Any help is appreciated.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p><code>$GITHUB_SHA</code> is shell syntax, but the <code>tags</code> argument isn’t a shell. Pretty much the only place in an Actions workflow where you have a shell are <code>run</code> steps.</p>
<p>You need to use Actions syntax here. I think <code>${{ github.sha }}</code> is what you want, see <a href="https://docs.github.com/en/actions/learn-github-actions/contexts#github-context"><code>github</code> context</a>.</p>

**Comment by kevinadhiguna** [comment]
<p>Thanks <a class="mention" href="/airtower-luna">@airtower-luna</a> !</p>
<p>It worked.</p>

**Comment by brianjmurrell** [comment]
But if one did want to do a concatenation in a shell `run` step such as:

```
echo "contains: ${{ contains(needs.Import-commit-pragmas.outputs.pr-repos, ${{ inputs.NAME }}@) }}"
```

what is the correct way to do that so that the second argument of `contains()` is `inputs.NAME` plus `@` concatenated?

And then alternatively what if one wanted to do the same in an `if:` statement such as:

```yaml
    if: |
        ${{ !contains(needs.Import-commit-pragmas.outputs.pr-repos, ${{inputs.NAME}}@) }}
```

**Reply by airtower-luna** [reply]
You'll probably need [the `format()` function](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#format), something similar to:

```
contains(needs.Import-commit-pragmas.outputs.pr-repos, format('{0}@', inputs.NAME))
```

Mind that using `${{ }}` templating in `run` scripts can create injection vulnerabilities, probably better to set an environment variable for the step and use that in the script.
